controller/src/lib.rs

```diff
@@ -254,13 +335,23 @@ impl Controller {
             message,
             data: None,
         };
-        let reply = self.rpc(request).await.unwrap();
+        let reply = self.rpc(request).await?;
+
+        // If we get back a ReadFailed error, one possibility is that we asked
+        // to read a transceiver that's not present.
+        let data = match reply.message.body {
+            MessageBody::SpResponse(SpResponse::Error(MessageError::ReadFailed)) => {
+                return Err(Error::ReadFailed);
+            }
+            MessageBody::SpResponse(SpResponse::Error(e)) => return Err(Error::from(e)),
+            MessageBody::SpResponse(SpResponse::Read(_)) => reply.data.unwrap(),
+            _ => panic!("Unexpected message type"),
+        };
+
         // We expect data to be a flattened vec of vecs, with the data from each
         // referenced transceiver. Split it into chunks sized by the number of
         // bytes we expected to read.
-        let data = reply
-            .data
-            .unwrap()
+        let data = data
             .chunks_exact(usize::from(read.len()))
             .map(Vec::from)
             .collect::<Vec<_>>();
```

Comment:
There's an assert just below this line in the diff and not new in the PR, but I have a similar question as on the panic above: does `self.rpc()` guarantee that `data.len()` is exactly what we expect in this assertion?

Reply:
Yes, `self.rpc()` guarantees that the length of the data is exactly what we expect. That's done using the `Message::expected_data_len` method. So this implicit panic was intentional.

Now that I think about it, one fallout of https://github.com/oxidecomputer/transceiver-control/issues/17 might be that this would need to change slightly. Short reads may become possible, if a module is unplugged while the request is being processed by the SP. It shouldn't change this part here, but the check for the expected data length should use the _response_ message to determine that, rather than the request.

Reply:
Actually, this already uses the response to determine how much data to expect. I think this is OK at this point.